code/game/machinery/vending.dm

```diff
@@ -605,6 +431,102 @@
 	else
 		return FALSE
 
+/obj/machinery/vending/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, \
+															datum/tgui/master_ui = null, datum/ui_state/state = GLOB.physical_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "vending", name, 350, 475, master_ui, state)
+		ui.open()
+
+/obj/machinery/vending/ui_data()
+	var/list/data = list()
+	var/list/listed_products = list()
+	var/list/display_records = product_records
+	if(extended_inventory)
+		display_records += hidden_records
+	if(coin)
+		display_records += coin_records
+	for(var/key = 1 to display_records.len)
+		var/datum/data/vending_product/I = display_records[key]
+		listed_products.Add(list(list(
+			"key" = key,
+			"name" = I.product_name,
+			"color" = I.display_color,
+			"amount" = I.amount)))
+	data["products"] = listed_products
+	if(!isnull(coin))
+		data["coin"] = coin.name
+	data["coinslot"] = premium.len
+	data["canvend"] = vend_ready
+
+	return data
+
+/obj/machinery/vending/ui_act(action, params)
+	if(..())
+		return
+	if(!vend_ready)
+		return
+	if(issilicon(usr))
+		if(iscyborg(usr))
+			var/mob/living/silicon/robot/R = usr
+			if(!(R.module && istype(R.module,/obj/item/robot_module/butler) ))
+				to_chat("<span class='notice'>[src] refuses to interface with you, as you are not in its target demographic!</span>")
+				return
+		else
+			to_chat(usr, "<span class='notice'>[src] refuses to interface with you, as you are not in its target demographic!</span>")
+			return
+	switch(action)
+		if("vend")
+			if(!allowed(usr) && !emagged && scan_id)	//For SECURE VENDING MACHINES YEAH
+				to_chat(usr, "<span class='warning'>Access denied.</span>")
+				flick(icon_deny,src)
+				return
+			vend_ready = FALSE
+			addtimer(VARSET_CALLBACK(src, vend_ready, TRUE), 10)
+			var/key = text2num(params["key"])
+			var/datum/data/vending_product/R = product_records[key]
+			if(R in hidden_records)
+				if(!extended_inventory)
+					return
+			else if(R in coin_records)
+				if(!coin)
+					to_chat(usr, "<span class='warning'>You need to insert a coin to get this item!</span>")
+					return
+				if(coin.string_attached)
+					if(prob(50))
+						if(usr.put_in_hands(coin))
+							to_chat(usr, "<span class='notice'>You successfully pull [coin] out before [src] could swallow it.</span>")
+							coin = null
+						else
+							to_chat(usr, "<span class='warning'>You couldn't pull [coin] out because your hands are full!</span>")
+							QDEL_NULL(coin)
+					else
+						to_chat(usr, "<span class='warning'>You weren't able to pull [coin] out fast enough, the machine ate it, string and all!</span>")
+						QDEL_NULL(coin)
+				else
+					QDEL_NULL(coin)
+			else if (!(R in product_records))

```

Comment:
you're missing the checks for coin records and hidden records

Reply:
that's not needed, as coin/hidden records have already been handled